### *“People ask me all the time: ‘What is it like to be a woman at Google?’* 

### *I’m not a woman at Google, I’m a geek at Google. And being a geek is just great. I’m a geek, I like to code, I even like to use spreadsheets when I cook.”* 

### **-Marissa Mayer**

----

## YOUR NAME HERE: Cody Turpin

----

# Homework 3: Reviewing Tables/DataFrames

Covers material from Ch 6 in the text.

In [3]:
#Please import the 2 packages you need for working with tables and data frames here
import pandas as pd
import numpy as np

## A. Differences between Universities

Suppose you're choosing a university to attend, and you'd like to *quantify* how *dissimilar* any two universities are.  You rate each university you're considering on several numerical traits.  You decide on a very detailed list of 1000 traits, and you measure all of them!  Some examples:

* The cost to attend (per year)
* The average Yelp review of nearby Thai restaurants
* The USA Today ranking of the Medical school
* The USA Today ranking of the Engineering school

You decide that the dissimilarity between two universities is the *total* of the differences in their traits.  That is, the dissimilarity is:

* the **sum** of
* the absolute values of
* the 1000 differences in their trait values.

### **Question 1.** 

A) In the next cell, load the School1.csv\School2.csv data, which contain 1000 trait values for 2 schools. 

In [4]:
school1 = pd.read_csv("School1.csv")
school2 = pd.read_csv('School2.csv') #Read in School2.csv

B)  Compute the dissimilarity (according to the  method described above) between School1 and School2.  

Call your answer `dissimilarity`.  Use a single line of code to compute the dissimilarity answer.

In [4]:
dissimilarity = sum(abs(school1['Trait value']-school2['Trait value']))
dissimilarity

14060.55870106792

Double-click __here__ for the correct answer.

<!-- 14060.55870...
-->

### **Question 2.** 

Why do we sum up the absolute values of the differences in trait values, rather than just summing up the differences?

Because you want to quantify the difference between the values in the two data sets which means you want to find the distance between the two. Distance can't be negative and this is the same case in trying to quanitfy the differences between the two sets so an absolute value should be taken.

---

##### Weighing the traits
After computing dissimilarities between several schools, you notice a problem with your method: the scale of the traits matters a lot.

Since schools cost tens of thousands of dollars to attend, the cost-to-attend trait is always a much bigger *number* than most other traits.  That makes it affect the dissimilarity a lot more than other traits.  Two schools that differ in cost-to-attend by \\$900, but are otherwise identical, get a dissimilarity of 900.  But two schools that differ in graduation rate by .9 (a huge difference!), but are otherwise identical, get a dissimilarity of only .9.

One way to fix this problem is to assign different "weights" to different traits.  For example, we could fix the problem above by multiplying the difference in the cost-to-attend traits by .001, so that a difference of \\$900 in cost-to-attend results in a dissimilarity of $ 900 \times .001 $, or $ .9 $.

Here's a revised method that does that for every trait:

1. For each trait, subtract the two schools' trait values.
2. Then take the absolute value of that difference.
3. *Now multiply that absolute value by a trait-specific number, like .001 or 2.*
4. Now sum the 1000 resulting numbers.

### **Question 3.** 

Suppose you've already decided on a weight for each trait. You saved the weights in a ``weights.csv`` file. 

Load them into a table called `weights` in the cell below. 

In [5]:
weights = pd.read_csv('weights.csv')
weights

Weight
0     0.001
1     0.200
2     0.050
3     0.050
4     0.200
..      ...
995   0.050
996   0.050
997   0.050
998   0.050
999   0.200

[1000 rows x 1 columns]

---

### **Question 4.**  

Now use the revised method to compute a revised dissimilarity between School 1 and School 2.

*Hint:* Using array arithmetic, your answer should be almost as short as in question 1.

In [6]:
revised_dissimilarity =  sum(abs(school1['Trait value']-school2['Trait value'])*weights['Weight'])
revised_dissimilarity

505.98313211458793

Double-click __here__ for the correct answer.

<!-- 505.98313211...
-->

---

## B. Little Women

Recall from our first week that we considered a plot that compares the number of characters to the number of periods in each chapter of Little Women.  Each point represents one chapter.

<img src="https://github.com/kathleen-ryan-DeSales/CS250/blob/main/pictures/little_women.png?raw=true">

### **Question 5.**

To generate the above graph, we started with the data in the table below.  

Show the first 10 rows using the ``head`` function so that you can succintly see what's in the table.

In [7]:
lw_df = pd.read_csv("little_women.csv")  
lw_df.head(10)

#Your code here

Chapter  Chapter Length  Number of Periods
0        1           23445                189
1        2           23785                188
2        3           22414                231
3        4           27272                195
4        5           25438                255
5        6           15639                140
6        7           15522                131
7        8           24174                214
8        9           36364                337
9       10           21196                185

### **Question 6.**

To generate the above graph, we need to compute the number of characters per period in each chapter.  Add a column to the ``lw_df`` called ``Chars Per Period`` with this information.

In [8]:
... #Your code here
lw_df['Chars Per Period'] = np.array(lw_df['Chapter Length'] / lw_df['Number of Periods'])
lw_df  #Print the table out and see your change!

Chapter  Chapter Length  Number of Periods  Chars Per Period
0         1           23445                189        124.047619
1         2           23785                188        126.515957
2         3           22414                231         97.030303
3         4           27272                195        139.856410
4         5           25438                255         99.756863
5         6           15639                140        111.707143
6         7           15522                131        118.488550
7         8           24174                214        112.962617
8         9           36364                337        107.905045
9        10           21196                185        114.572973
10       11           24745                193        128.212435
11       12           42549                429         99.181818
12       13           20208                175        115.474286
13       14           19623                180        109.016667
14       15           18785                181        103.784530
15       16           16688                172         97.023256
16       17           15129                155         97.606452
17       18           18651                144        129.520833
18       19           18538                121        153.206612
19       20           15327                145        105.703448
20       21           25550                269         94.981413
21       22           14849                120        123.741667
22       23           24889                247        100.765182
23       24           25976                182        142.725275
24       25           14455                 91        158.846154
25       26           23267                150        155.113333
26       27           16781                109        153.954128
27       28           33147                271        122.313653
28       29           28239                233        121.197425
29       30           25899                218        118.802752
30       31           22180                178        124.606742
31       32           25058                224        111.866071
32       33           27580                232        118.879310
33       34           34587                257        134.579767
34       35           23014                201        114.497512
35       36           12245                100        122.450000
36       37           25171                157        160.324841
37       38           27986                206        135.854369
38       39           29715                263        112.984791
39       40           12335                 61        202.213115
40       41           28098                187        150.256684
41       42           17882                118        151.542373
42       43           35648                305        116.878689
43       44           11106                 95        116.905263
44       45           13462                 96        140.229167
45       46           29623                234        126.594017
46       47           43852                392        111.867347

### **Question 7.**

Find highest proportion of characrters per period of all chapters.

In [9]:
... #More code here

most_chars_per_period = max(lw_df['Chars Per Period'])
most_chars_per_period  #This variable should contain your inal answer

202.21311475409837

### **Question 8.**

Without hard coding, find the chapter with the lowest proportion of characters per period.

In [10]:
... #More code here

chapter_least_chars_per_period = (lw_df.sort_values(by='Chars Per Period').iloc[0])
chapter_least_chars_per_period  #This variable should contain your final answer

Chapter                 21.000000
Chapter Length       25550.000000
Number of Periods      269.000000
Chars Per Period        94.981413
Name: 20, dtype: float64

### **Question 9.**

Find the average and standard deviation of the characters per period for the chapters in little women.

In [11]:
mean_chars_per_period = lw_df['Chars Per Period'].mean()
std_chars_per_period = lw_df['Chars Per Period'].std()

mean_chars_per_period, std_chars_per_period  #These variables should contain your final answer

(124.181805439536, 21.416193543551703)

---

## C. World Populations


Let's Review basic ``where`` clauses.


### **Question 10.**

a) Read in the world_population.csv below. The table has a single column with the world population in each year from 1950-2015 (as [estimated](http://www.census.gov/population/international/data/worldpop/table_population.php) by the US Census Bureau).

In [5]:
#Read in table
population_table = pd.read_csv('world_population.csv')
population_table

Population
0   2557628654
1   2594939877
2   2636772306
3   2682053389
4   2730228104
..         ...
61  6944055583
62  7022349283
63  7101027895
64  7178722893
65  7256490011

[66 rows x 1 columns]

### **Question 11.**

Using ``arange``, create a column called "Year" in the table that shows the years.  The first row is 1950 and the last row is 2015.

In [6]:
#Create the Years Column here
population_table['Year'] = np.arange(1950, 2016)

population_table

Population  Year
0   2557628654  1950
1   2594939877  1951
2   2636772306  1952
3   2682053389  1953
4   2730228104  1954
..         ...   ...
61  6944055583  2011
62  7022349283  2012
63  7101027895  2013
64  7178722893  2014
65  7256490011  2015

[66 rows x 2 columns]

### **Question 12.**

Use boolean indexing (so create a "where" clause) to answer this question:  In what year did world population cross 6 billion?

In [7]:
#do some work first - Add as many lines/cells as you need

#To get credit, this  variable should contain your 
#FINAL ANSWER so don't use it for other variable names

where_6bil = (population_table['Population'] >= 6000000000)
where_6bil = population_table[where_6bil]
where_6bil
yr_over_6bil = (where_6bil.sort_values(by='Population').iloc[0])
yr_over_6bil

Population    6012074922
Year                1999
Name: 49, dtype: int64

---

## D. Creating your own csv of data

Let's load in data that you yourself have created.


### **Question 13.**

Create your own CSV file called `my_data_yourlastname.txt` inside the same folder as this lab's folder.  Your file have **two columns** and **four rows**, and one of the rows should contain column headers. It can have whatever values you want.  

NOTE:  

* Even though it is marked as a `.txt` file, you can just add commas to designate columns and it will be loaded in like a csv. 

* To create the file, go to the folder containing this lab and right click to select "New -> Text File".  Then rename the file `my_data.txt`.  

Below, load the file into a dataframe called `my_data`.

In [8]:
# Load your table here and make sure it's printed out without an error!
my_data = pd.read_csv('my_data_turpin.txt')
my_data

Subject\tGrades
0        68\tMath
1     92\tFinance
2     87\tScience

---

## E. Creating tables from scratch within Jupyter Notebooks

Let's review creating tables from scratch from within a notebook file.

### **Question 14.**

In the cell below, create a table `top_10_movies` with 2 columns: "Rating" and "Name" which have the data below.  These data corresponds to the [IMDb](imdb.com) ratings of 8 movies.  
* 9.2, 9.2, 9., 8.9, 8.9, 8.9, 8.9, 8.9, 8.9, 8.8
* These movies:
        'The Shawshank Redemption (1994)',
        'The Godfather (1972)',
        'The Godfather: Part II (1974)',
        'Pulp Fiction (1994)',
        "Schindler's List (1993)",
        'The Lord of the Rings: The Return of the King (2003)',
        '12 Angry Men (1957)',
        'The Dark Knight (2008)',
        'Il buono, il brutto, il cattivo (1966)',
        'The Lord of the Rings: The Fellowship of the Ring (2001)'


In [16]:
#Add as many lines as you need to do this.

top_10_movies_dict = {'Rating' : np.array(['9.2', '9.2', '9.', '8.9', '8.9', '8.9', '8.9', '8.9', '8.9', '8.8']),
                 'Name' : np.array(['The Shawshank Redemption (1994)',
  'The Godfather (1972)',
  'The Godfather: Part II (1974)',
  'Pulp Fiction (1994)',
  "Schindler's List (1993)",
  'The Lord of the Rings: The Return of the King (2003)',
  '12 Angry Men (1957)',
  'The Dark Knight (2008)',
  'Il buono, il brutto, il cattivo (1966)',
  'The Lord of the Rings: The Fellowship of the Ring (2001)'])}

top_10_movies = pd.DataFrame(top_10_movies_dict)  #Should contain your table
top_10_movies

Rating                                               Name
0    9.2                    The Shawshank Redemption (1994)
1    9.2                               The Godfather (1972)
2     9.                      The Godfather: Part II (1974)
3    8.9                                Pulp Fiction (1994)
4    8.9                            Schindler's List (1993)
5    8.9  The Lord of the Rings: The Return of the King ...
6    8.9                                12 Angry Men (1957)
7    8.9                             The Dark Knight (2008)
8    8.9             Il buono, il brutto, il cattivo (1966)
9    8.8  The Lord of the Rings: The Fellowship of the R...

---

## F. IMDb Data

Now let's load in more IMDb data!

### **Question 15.**

a) The file `imdb.csv` contains a table of information about the 250 highest-rated movies on IMDb.  Load it as a table called `imdb`.

In [9]:
#Read the file in here.
imdb = pd.read_csv('imdb.csv')
imdb

Votes  Rating                                         Title  Year
0      88355     8.4                                      M (1931)  1931
1     132823     8.3                    Singin' in the Rain (1952)  1952
2      74178     8.3                          All About Eve (1950)  1950
3     635139     8.6                                   Léon (1994)  1994
4     145514     8.2                       The Elephant Man (1980)  1980
..       ...     ...                                           ...   ...
245  1078416     8.7                           Forrest Gump (1994)  1994
246    31003     8.1                  Le salaire de la peur (1953)  1953
247   167076     8.2                               3 Idiots (2009)  2009
248    91689     8.1                                Network (1976)  1976
249   589477     8.3  Eternal Sunshine of the Spotless Mind (2004)  2004

[250 rows x 4 columns]

Notice this table is big enough that only a few of its rows are displayed, but the others are still there.  10 are shown, so there are 250 movies total.

### **Question 16.**

In the next cell, find the highest rating in the database **without** hardcoding.

In [12]:
#Add lines here

highest_rating = imdb.max()
highest_rating #Should contain your FINAL ANSWER

Votes            1498733
Rating               9.2
Title     Yôjinbô (1961)
Year                2015
dtype: object

### **Question 17.**

Now find how many movies had the highest rating **without** hardcoding.

In [14]:
#Add as many lines/cells as you need
num_with_highest_rating = (imdb['Rating'] == highest_rating['Rating'])
num_with_highest_rating = imdb[num_with_highest_rating]

#Don't hardcode!
num_with_highest_rating  #Must contain your final answer

Votes  Rating                            Title  Year
22  1027398     9.2             The Godfather (1972)  1972
53  1498733     9.2  The Shawshank Redemption (1994)  1994

### **Question 18.**

In the cell below, sort your data from highest to lowest, as to confirm your answer above.

In [20]:
#Your code here
imdb.sort_values(by='Rating', ascending=False)

Votes  Rating                                   Title  Year
22   1027398     9.2                    The Godfather (1972)  1972
53   1498733     9.2         The Shawshank Redemption (1994)  1994
91    692753     9.0           The Godfather: Part II (1974)  1974
105   384187     8.9                     12 Angry Men (1957)  1957
57    447875     8.9  Il buono, il brutto, il cattivo (1966)  1966
..       ...     ...                                     ...   ...
168   500576     8.0                   Monsters, Inc. (2001)  2001
166    59578     8.0                    The Big Sleep (1946)  1946
46    427099     8.0       X-Men: Days of Future Past (2014)  2014
51     87437     8.0                    Roman Holiday (1953)  1953
125    26012     8.0                      Kumonosu-jô (1957)  1957

[250 rows x 4 columns]

### **Question 19.**

Now find all movie names **(so just the movie names and no other details)** with the LOWEST rating.  

In [21]:
lowest_rating = imdb.min()  #HINT:  First find the lowest rating
print(lowest_rating)

#Then find the movies with that lowest rating:
lowest_titles = (imdb['Rating'] == lowest_rating['Rating'])
lowest_titles = imdb[lowest_titles]
lowest_titles

Votes                   26012
Rating                    8.0
Title     12 Angry Men (1957)
Year                     1921
dtype: object


Votes  Rating                                              Title  Year
9     46987     8.0                            Relatos salvajes (2014)  2014
23    55382     8.0        Bom yeoreum gaeul gyeoul geurigo bom (2003)  2003
28    32385     8.0                      La battaglia di Algeri (1966)  1966
31   364225     8.0                                        Jaws (1975)  1975
32   158867     8.0                              Before Sunrise (1995)  1995
35    56671     8.0                                 The Killing (1956)  1956
36    87591     8.0                                    Papillon (1973)  1973
43    43090     8.0                                Paris, Texas (1984)  1984
46   427099     8.0                  X-Men: Days of Future Past (2014)  2014
51    87437     8.0                               Roman Holiday (1953)  1953
56   321834     8.0                          The Imitation Game (2014)  2014
71    89920     8.0                                Barry Lyndon (1975)  1975
77    42446     8.0                                   La strada (1954)  1954
85   268480     8.0                        Beauty and the Beast (1991)  1991
92    49135     8.0            The Man Who Shot Liberty Valance (1962)  1962
93   527349     8.0                     Guardians of the Galaxy (2014)  2014
108  466111     8.0  Harry Potter and the Deathly Hallows: Part 2 (...  2011
111   54651     8.0                              Salinui chueok (2003)  2003
112   82440     8.0                                Mou gaan dou (2002)  2002
116   90937     8.0                                    La haine (1995)  1995
117  718645     8.0  Pirates of the Caribbean: The Curse of the Bla...  2003
120   66622     8.0                                   Notorious (1946)  1946
122  138240     8.0                                     Yip Man (2008)  2008
123  384272     8.0                               Groundhog Day (1993)  1993
124   91652     8.0                                       Akira (1988)  1988
125   26012     8.0                                 Kumonosu-jô (1957)  1957
126  415809     8.0                              Twelve Monkeys (1995)  1995
127  527274     8.0                              The Terminator (1984)  1984
128   50208     8.0             Who's Afraid of Virginia Woolf? (1966)  1966
131  282774     8.0                                    The Help (2011)  2011
132   37972     8.0                         Fanny och Alexander (1982)  1982
134   57644     8.0                       Trois couleurs: Rouge (1994)  1994
144   59377     8.0           Lagaan: Once Upon a Time in India (2001)  2001
154  520391     8.0                               Jurassic Park (1993)  1993
166   59578     8.0                               The Big Sleep (1946)  1946
168  500576     8.0                              Monsters, Inc. (2001)  2001
172  147277     8.0                               Before Sunset (2004)  2004
174  862016     8.0                                The Avengers (2012)  2012
176  124671     8.0                     Per un pugno di dollari (1964)  1964
180   39447     8.0                                 Underground (1995)  1995
182   28012     8.0                                 Le samouraï (1967)  1967
191  434906     8.0                           The King's Speech (2010)  2010
192   72007     8.0                                   High Noon (1952)  1952
193  312516     8.0                                   Prisoners (2013)  2013
196  164419     8.0                           Dog Day Afternoon (1975)  1975
206  583004     8.0                             The Truman Show (1998)  1998
214   57842     8.0                                 The Hustler (1961)  1961
221  318041     8.0                                       Rocky (1976)  1976
228   83941     8.0                      Tenkû no shiro Rapyuta (1986)  1986
237   57974     8.0                     The Night of the Hunter (1955)  1955

e) Now create a version of `imdb` that's sorted chronologically, with the earliest movies first.  Call it `imdb_by_year`.

In [22]:
#Your code here
imdb_by_year = imdb.sort_values(by='Year')
imdb_by_year #print it out to see!

Votes  Rating                            Title  Year
173   55784     8.3                   The Kid (1921)  1921
205   58506     8.2             The Gold Rush (1925)  1925
146   46332     8.2               The General (1926)  1926
49    98794     8.3                Metropolis (1927)  1927
0     88355     8.4                         M (1931)  1931
..      ...     ...                              ...   ...
100  369141     8.1  The Grand Budapest Hotel (2014)  2014
9     46987     8.0          Relatos salvajes (2014)  2014
70   689541     8.6              Interstellar (2014)  2014
233  262425     8.3        Mad Max: Fury Road (2015)  2015
41    79615     8.5              Inside Out (2015/I)  2015

[250 rows x 4 columns]

### **Question 20.**

What's the title of the earliest movie in the dataset? Write code in the cell below to decide.  

**Do not hardcode.**

**Show just the title and no other info.**

In [23]:
earliest_movie_title = imdb_by_year.iloc[0] #Don't hardcode!
earliest_movie_title

Votes              55784
Rating               8.3
Title     The Kid (1921)
Year                1921
Name: 173, dtype: object

---

You're ready for more!

---

### **Question 21.**

Suppose you're interested in movies from the 1940s.  Set ``forties`` to a subtable containing only the movies from 1940s.

In [24]:
#Your code here - Add as many lines/cells as you need
forties = (imdb['Year'] > 1939) & (imdb['Year'] < 1950)
forties = imdb[forties]



forties #Should print out your final answer

Votes  Rating                                    Title  Year
21    55793     8.1               The Grapes of Wrath (1940)  1940
50    86715     8.3                  Double Indemnity (1944)  1944
72   101754     8.1                The Maltese Falcon (1941)  1941
75    71003     8.3  The Treasure of the Sierra Madre (1948)  1948
102   35983     8.1       The Best Years of Our Lives (1946)  1946
118   81887     8.3               Ladri di biciclette (1948)  1948
120   66622     8.0                         Notorious (1946)  1946
158  350551     8.5                        Casablanca (1942)  1942
166   59578     8.0                     The Big Sleep (1946)  1946
167   78216     8.2                           Rebecca (1940)  1940
213  242353     8.6             It's a Wonderful Life (1946)  1946
215  101785     8.3                     The Third Man (1949)  1949
240  117590     8.4                The Great Dictator (1940)  1940
243  268905     8.4                      Citizen Kane (1941)  1941

### **Question 22.**

Create a table called `nineties` containing the movies that came out in the 1990s. 

In [25]:
#Your code here - Add as many lines/cells as you need
nineties = (imdb['Year'] > 1989) & (imdb['Year'] < 2000)
nineties = imdb[nineties]



nineties #SHould print out your final answer

Votes  Rating                                       Title  Year
3     635139     8.6                                 Léon (1994)  1994
15    192165     8.4                        Mononoke-hime (1997)  1997
16    769893     8.5                  Saving Private Ryan (1998)  1998
20     95212     8.1            In the Name of the Father (1993)  1993
32    158867     8.0                       Before Sunrise (1995)  1995
47    767224     8.6             The Silence of the Lambs (1991)  1991
53   1498733     9.2             The Shawshank Redemption (1994)  1994
66    388239     8.2                                 Heat (1995)  1995
79    535249     8.3                            Toy Story (1995)  1995
85    268480     8.0                 Beauty and the Beast (1991)  1991
87   1177098     8.8                           Fight Club (1999)  1999
94    376590     8.3                    L.A. Confidential (1997)  1997
96    656756     8.6                   The Usual Suspects (1995)  1995
97    658564     8.5           Terminator 2: Judgment Day (1991)  1991
103   644556     8.7                           Goodfellas (1990)  1990
104   735056     8.4                      American Beauty (1999)  1999
115   630994     8.1                      The Sixth Sense (1999)  1999
116    90937     8.0                             La haine (1995)  1995
123   384272     8.0                        Groundhog Day (1993)  1993
126   415809     8.0                       Twelve Monkeys (1995)  1995
129  1073043     8.7                           The Matrix (1999)  1999
130   248514     8.3                           Unforgiven (1992)  1992
133   694602     8.5                   American History X (1998)  1998
134    57644     8.0                Trois couleurs: Rouge (1994)  1994
138   578684     8.3                       Reservoir Dogs (1992)  1992
139   294394     8.2                               Casino (1995)  1995
147   761224     8.9                     Schindler's List (1993)  1993
149   672878     8.5                       The Green Mile (1999)  1999
154   520391     8.0                        Jurassic Park (1993)  1993
155   372863     8.2  Lock, Stock and Two Smoking Barrels (1998)  1998
159   395997     8.1                                Fargo (1996)  1996
161   529800     8.2                    Good Will Hunting (1997)  1997
163   419372     8.1                        Trainspotting (1996)  1996
169   895411     8.6                                Se7en (1995)  1995
170   653769     8.3                           Braveheart (1995)  1995
180    39447     8.0                          Underground (1995)  1995
201   473988     8.2                     The Big Lebowski (1998)  1998
202   548750     8.4                        The Lion King (1994)  1994
204   358305     8.6                      La vita è bella (1997)  1997
206   583004     8.0                      The Truman Show (1998)  1998
224  1166532     8.9                         Pulp Fiction (1994)  1994
245  1078416     8.7                         Forrest Gump (1994)  1994

### **Question 23.**

How does the average and standard deviation of the movie ratings from the 40s and 90s compare? 

Use the ``describe`` function to help you assess this quickly.

The syntax for ``describe`` is as you would expect:   ``table["column_name"].describe().``

In [26]:
#Call describe on forties here.
forties.describe()

Votes     Rating         Year
count      14.000000  14.000000    14.000000
mean   122766.785714   8.257143  1944.071429
std     94155.592749   0.182775     3.315796
min     35983.000000   8.000000  1940.000000
25%     67717.250000   8.100000  1941.000000
50%     84301.000000   8.300000  1945.000000
75%    113638.750000   8.375000  1946.000000
max    350551.000000   8.600000  1949.000000

In [27]:
#Call describe on nineties here.
nineties.describe()

Votes     Rating         Year
count  4.200000e+01  42.000000    42.000000
mean   5.594491e+05   8.357143  1995.190476
std    3.225620e+05   0.303751     2.491332
min    3.944700e+04   8.000000  1990.000000
25%    3.737948e+05   8.100000  1994.000000
50%    5.419995e+05   8.300000  1995.000000
75%    6.891710e+05   8.600000  1997.000000
max    1.498733e+06   9.200000  1999.000000

Looking at the output of describe, answer below these 3 questions:

    A. Did the 90s or 40s have a higher average rating?  
    
    B. You should see that the 90s movie ratings are more spread out from the mean on average?  How am I making this decision ased on the results of the describe function?
    
    C. Why is it sensible that the movie ratings in the 90s would show more variation from the mean than from those in the 40s?  Give a reason based on your world view.

A.  Your Answer Here: The nineties has a slightly higher average rating (8.35)

B.  Your Answer Here: The standard deviation for the ratings in the nineties is larger (.303 vs .182) than the forties meaning there is more spread in the ratings

C.  Your Answer Here: The nineties had a lot more critics than the forties did simply because movies overall were much more popular in the nineties than in the forties and had decades of comparable movies to go off of. Also, there were far more movies made during the nineties compared to the forties which could be assumed that the raters in the forties couldn't give bad ratings too often compared to the nineties due to selection size.


---

### **Question 24.**

Now find all the movies with more than 1 million votes on IMDb.

In [28]:
#You got this!  Add as many lines as you need
movies_lots_of_votes = (imdb['Votes'] > 1000000)

movies_lots_of_votes = imdb[movies_lots_of_votes]
movies_lots_of_votes #Should print out your final answer

Votes  Rating                                              Title  Year
22   1027398     9.2                               The Godfather (1972)  1972
53   1498733     9.2                    The Shawshank Redemption (1994)  1994
76   1473049     8.9                             The Dark Knight (2008)  2008
81   1271949     8.7                                   Inception (2010)  2010
87   1177098     8.8                                  Fight Club (1999)  1999
129  1073043     8.7                                  The Matrix (1999)  1999
141  1074146     8.9  The Lord of the Rings: The Return of the King ...  2003
178  1099087     8.8  The Lord of the Rings: The Fellowship of the R...  2001
224  1166532     8.9                                Pulp Fiction (1994)  1994
245  1078416     8.7                                Forrest Gump (1994)  1994

### **Question 25.**

Now find how many movies there were with a rating higher than 8.5.  

Start by creating a table called really_highly_rated containing all movies with a rating higher than 8.5

In [29]:
#Add as many lines/cells here

really_highly_rated = imdb['Rating'] > 8.5
really_highly_rated = imdb[really_highly_rated]

num_really_highly_rated = len(really_highly_rated)
num_really_highly_rated  #Should print out final answer

29

### **Question 26.**

Find the average rating for movies released in the 20th century and the average rating for movies released in the 21st century.

**Note:** Our `imdb` dataset includes 250 of the best-rated movies ever made.  There are millions of movies, and most are not represented in this dataset.  So whatever you find will be true only among these, not among movies in general.

In [30]:
where_20th_century = imdb['Year'] < 2000
where_20th_century = imdb[where_20th_century]
average_20th_century_rating = where_20th_century['Rating'].mean()

where_21st_century = imdb['Year'] > 1999
where_21st_century = imdb[where_21st_century]
average_21st_century_rating = where_21st_century['Rating'].mean()

print("Average 20th century rating:", average_20th_century_rating)
print("Average 21st century rating:", average_21st_century_rating)

Average 20th century rating: 8.278362573099418
Average 21st century rating: 8.237974683544303


### **Question 27.**

Below, determine how many rows and columns there are in the imdb dataset without hardcoding.

In [31]:
num_movies_in_dataset = len(imdb) #Don't hardcode
num_columns_in_dataset =  len(imdb.columns) #Don't hardcode
print("Number of Movies:", num_movies_in_dataset, "Number of Columns:", num_columns_in_dataset)

Number of Movies: 250 Number of Columns: 4


### **Question 28.**

Now find the *proportion* of movies in the dataset that were released in the 20th century, and the proportion from the 21st century.

*Hint:* The *proportion* is a synonym for *percentage* but often we leave the proportion as number between 0 and 1 (so 0.15 instead of 15%.)

*Hint:* If you do this correctly, you should get:

    Proportion in 20th century: 0.684
    Proportion in 21st century: 0.316

In [32]:
#Add code here if you need
proportion_20th = len(where_20th_century)
proportion_21st = len(where_21st_century)

proportion_in_20th_century = proportion_20th / num_movies_in_dataset
proportion_in_21st_century = proportion_21st / num_movies_in_dataset

print("Proportion in 20th century:", proportion_in_20th_century)
print("Proportion in 21st century:", proportion_in_21st_century)

Proportion in 20th century: 0.684
Proportion in 21st century: 0.316


---

### **Question 29.**

a) Here's a challenge: Find the number of movies that came out in *even* years.

*Hint:* The operator `%` computes the remainder when dividing by a number.  So `5 % 2` is 1 and `6 % 2` is 0.  A number is even if the remainder is 0 when you divide by 2.

*Hint 2:* `%` can be used on arrays, operating elementwise like `+` or `*`.  So `make_array(5, 6, 7) % 2` is `array([1, 0, 1])`.

*Hint 3:* Create a column called "Year Remainder" that's the remainder when each movie's release year is divided by 2.  Then use boolean indexing to find rows where that new column is equal to 0. 

In [33]:
# You can approach this however you like but you'll need some steps before you set the num_even_year_movies variable.
imdb ['Year Remainder'] = imdb['Year'] % 2
num_even = imdb['Year Remainder'] == 1
num_even = imdb[num_even]
num_even_year_movies = len(num_even)
num_even_year_movies

123

b) Use the ``not`` operator (~) to find the mean now select all odd years and find the mean rating of these movies.

In [34]:
# You must use ~ to get points for doing this.
imdb ['Year Remainder'] = imdb['Year'] % 2
num_odd = ~(imdb['Year Remainder'] == 1)
num_odd = imdb[num_odd]
odd_year_movies_mean = len(num_odd)
odd_year_movies_mean

127

---


### **Question 30.**

There are 7 movies that contain the word "man".  

Recall: Look at this [this website](https://www.tutorialspoint.com/python_pandas/python_pandas_working_with_text_data.htm) for  ``str`` functions that could help.

Your answer should appear as:

|     | Votes  | Rating | Title                                   | Year |
|-----|--------|--------|-----------------------------------------|------|
|   4 | 145514 |    8.2 |                 The Elephant Man (1980) | 1980 |
|   7 | 850601 |    8.3 |                    Batman Begins (2005) | 2005 |
|  51 |  87437 |    8.0 |                    Roman Holiday (1953) | 1953 |
|  92 |  49135 |    8.0 | The Man Who Shot Liberty Valance (1962) | 1962 |
| 122 | 138240 |    8.0 |                          Yip Man (2008) | 2008 |
| 206 | 583004 |    8.0 |                  The Truman Show (1998) | 1998 |
| 215 | 101785 |    8.3 |                    The Third Man (1949) | 1949 |

In [35]:
#Add as much code as you need here
Man = imdb['Title'].str.contains('Man')
Man = imdb[Man]
man = imdb['Title'].str.contains('man')
man = imdb[man]
Men = Man.append(man)
movies_with_man = Men
movies_with_man

Votes  Rating                                    Title  Year  \
4    145514     8.2                  The Elephant Man (1980)  1980   
92    49135     8.0  The Man Who Shot Liberty Valance (1962)  1962   
122  138240     8.0                           Yip Man (2008)  2008   
215  101785     8.3                     The Third Man (1949)  1949   
7    850601     8.3                     Batman Begins (2005)  2005   
51    87437     8.0                     Roman Holiday (1953)  1953   
206  583004     8.0                   The Truman Show (1998)  1998   

     Year Remainder  
4                 0  
92                0  
122               0  
215               1  
7                 1  
51                1  
206               0

---

## G. Farmers Markets

Let's change gears and work with a new data set.

The table `farmers_markets.csv` contains data on farmers' markets in the United States.  (The data are collected [by the USDA]([dataset](https://apps.ams.usda.gov/FarmersMarketsExport/ExcelExport.aspx)).)  Each row represents one such market.

You'll notice that it has a large number of columns in it!

Load the dataset into a table.  Call it `farmers_markets`.

In [36]:
farmers_markets = pd.read_csv("farmers_markets.csv")
farmers_markets

FMID                                        MarketName  \
0     1012063   Caledonia Farmers Market Association - Danville   
1     1011871                 Stearns Homestead Farmers' Market   
2     1011878                                   100 Mile Market   
3     1009364                 106 S. Main Street Farmers Market   
4     1010691               10th Steet Community Farmers Market   
...       ...                                               ...   
8541  1004767                    Zia Bernalillo Farmers' Market   
8542  1000778                          Zimmerman Farmers Market   
8543  1002868                        Zion Canyon Farmers Market   
8544  1004686                         Zionsville Farmers Market   
8545  1011418                         Zona Rosa Farmers' Market   

                                                Website  \
0     https://sites.google.com/site/caledoniafarmers...   
1                           http://Stearnshomestead.com   
2                             http://www.pfcmarkets.com   
3                http://thetownofsixmile.wordpress.com/   
4                                                   NaN   
...                                                 ...   
8541                         http://www.eatfreshnm.org/   
8542                     http://www.pzfarmersmarket.org   
8543                         http://www.zionharvest.org   
8544             http://www.zionsvillefarmersmarket.org   
8545                                                NaN   

                                               Facebook Twitter Youtube  \
0     https://www.facebook.com/Danville.VT.Farmers.M...     NaN     NaN   
1                                                   NaN     NaN     NaN   
2       https://www.facebook.com/100MileMarket/?fref=ts     NaN     NaN   
3                                                   NaN     NaN     NaN   
4                                                   NaN     NaN     NaN   
...                                                 ...     ...     ...   
8541   https://www.facebook.com/farmersmarketsnm?ref=hl     NaN     NaN   
8542                  princeton-zimerman farmers market     NaN     NaN   
8543  https://www.facebook.com/ZionCanyonFarmersMark...     NaN     NaN   
8544                                                NaN     NaN     NaN   
8545                                                NaN     NaN     NaN   

                                             OtherMedia  \
0                                                   NaN   
1                                                   NaN   
2              https://www.instagram.com/100milemarket/   
3                                                   NaN   
4     http://agrimissouri.com/mo-grown/grodetail.php...   
...                                                 ...   
8541                                                NaN   
8542                                                NaN   
8543                                                NaN   
8544                                                NaN   
8545                                                NaN   

                          street         city      County  ... Coffee Beans  \
0                            NaN    Danville    Caledonia  ...      Y     Y   
1                6975 Ridge Road       Parma     Cuyahoga  ...      N     N   
2                507 Harrison St    Kalamazoo   Kalamazoo  ...      N     N   
3             106 S. Main Street     Six Mile         NaN  ...      N     N   
4         10th Street and Poplar       Lamar       Barton  ...      N     N   
...                          ...          ...         ...  ...    ...   ...   
8541    335 S. Camino del Pueblo   Bernalillo    Sandoval  ...      N     Y   
8542    Lions Park, Main Street     Zimmerman   Sherburne  ...      N     N   
8543        1212 Zion Park Blvd.   Springdale  Washington  ...      Y     N   
8544     Hawthorne & Main Street   Zionsville       Boone  ...      N     N   
8545  Off of I-2

### **Question 31.**

Find the number of rows and columns in the table.

In [37]:
num_farmers_markets_rows = len(farmers_markets)
num_farmers_markets_columns = len(farmers_markets.columns)
print("The table has", num_farmers_markets_columns, "columns and ", num_farmers_markets_rows, "rows in it!")

The table has 59 columns and  8546 rows in it!


### Learning How to "Whittle Down" The Columns

Most of the columns are about particular products -- whether the market sells tofu, pet food, etc.  If we're not interested in that stuff, it just makes the table difficult to read.  This comes up more than you might think.

In such situations, we can pare down the columns of a table to the ones we need.  If we just want to select one column, we would just use this notation:  ``imdb["Year"]``.  However, when we select more than one columns, we need to pass in the list of columns ``["Year", "Decade"]`` and this means we have double hard brackets. 

For example, the value of `imdb[["Year", "Decade"]]` is a table with only the years and decades of each movie in `imdb`.  

### **Question 32.**

Create a table from our new ``farmers_markets`` dataframe that has only the name, city, state, latitude ('y'), and longitude ('x') of each market.  Call that new table `farmers_markets_locations`.

In [38]:
farmers_markets_locations = farmers_markets[['MarketName', 'city', 'State', 'y', 'x']]
farmers_markets_locations

MarketName         city  \
0      Caledonia Farmers Market Association - Danville    Danville    
1                    Stearns Homestead Farmers' Market       Parma    
2                                      100 Mile Market    Kalamazoo   
3                    106 S. Main Street Farmers Market     Six Mile   
4                  10th Steet Community Farmers Market       Lamar    
...                                                ...          ...   
8541                    Zia Bernalillo Farmers' Market   Bernalillo   
8542                          Zimmerman Farmers Market    Zimmerman   
8543                        Zion Canyon Farmers Market   Springdale   
8544                         Zionsville Farmers Market   Zionsville   
8545                         Zona Rosa Farmers' Market  Kansas City   

               State          y           x  
0            Vermont  44.411013  -72.140305  
1               Ohio  41.375118  -81.728597  
2           Michigan  42.296024  -85.574887  
3     South Carolina  34.804200  -82.818700  
4           Missouri  37.495628  -94.274619  
...              ...        ...         ...  
8541      New Mexico  35.313704 -106.546840  
8542       Minnesota  45.437700  -93.585014  
8543            Utah  37.184508 -113.003311  
8544         Indiana  39.949100  -86.261200  
8545        Missouri  39.246317  -94.580864  

[8546 rows x 5 columns]

### **Question 33.**

Find the average latitude ('y') and average longitude ('x') of the farmers' markets in the table below.

In [39]:
average_latitude = farmers_markets['y'].mean()
average_longitude = farmers_markets['x'].mean()
print("The average of US farmers' markets' coordinates is located at (", average_latitude, ",", average_longitude, ")")

The average of US farmers' markets' coordinates is located at ( 39.18646452349536 , -90.99258081292632 )


### **Question 34.**

Make a table of the 5 northernmost farmers' markets in `farmers_markets_locations`.  Call it `northern_markets`.  It should include the same columns as `farmers_markets_locations`.

In [40]:
#Add more code here if you prefer to take more than one line to accomplish the task.

northern_markets = farmers_markets_locations.sort_values(by='y', ascending=False).iloc[0:5]
northern_markets

MarketName            city   State          y  \
7274   Tanana Valley Farmers Market       Fairbanks  Alaska  64.862750   
2392         Ester Community Market           Ester  Alaska  64.845900   
2430      Fairbanks Downtown Market       Fairbanks  Alaska  64.844414   
5182         Nenana Open Air Market          Nenana  Alaska  64.556600   
3496  Highway's End Farmers' Market  Delta Junction  Alaska  64.038462   

               x  
7274 -147.781100  
2392 -148.010000  
2430 -147.719593  
5182 -149.096000  
3496 -145.733115

### **Question 35.**

Make a table of the farmers' markets in Allentown, Pennyslvania.  **Do not assume that only PA has a city named Allentown when designing your where clause.** 

Your table should include the same columns as `farmers_markets_locations`.

In [41]:
#Add more code here if you prefer to take more than one line to accomplish the task.
allentown_markets = (farmers_markets_locations['State'] == 'Pennsylvania') & (farmers_markets_locations['city'] == 'Allentown')
allentown_markets = farmers_markets_locations[allentown_markets]
allentown_markets

MarketName       city         State  \
142     Allentown Farigrounds Farmers Market  Allentown  Pennsylvania   
143  Allentown Turnpike Plaza Farmers Market  Allentown  Pennsylvania   

           y        x  
142  40.6008 -75.4942  
143  40.6218 -79.7254

---

You're done!  Congrats!  

**Be sure to submit both of these files:**
* **the ``my_data_yourlastname.txt`` file** 
* **this ``ipynb`` file.**